### PR TITLE
Docs for KNIDEPLOY-3583

### PR DESCRIPTION
**Pull request by johnwilkins**
@ahardin-rh for [enterprise-4](https://issues.redhat.com/browse/enterprise-4).7. Still needs a QE review.

https://issues.redhat.com/browse/KNIDEPLOY-3583

Fixes:

https://issues.redhat.com/browse/TELCODOCS-117
https://issues.redhat.com/browse/TELCODOCS-123
https://issues.redhat.com/browse/TELCODOCS-137

### CODE REVIEW

File: modules/k8s-nmstate-creating-interfaces.adoc [hunk 1]
```diff
@@ -0,0 +1,250 @@
+// This is included in the following assemblies:
+//
+// k8s-nmstate-configuration.adoc
```
Comment (ahardin-rh) [inline modules/k8s-nmstate-creating-interfaces.adoc:3]: List the full path to the assembly: 

    * networking/k8s_nmstate/k8s-nmstate-configuring.adoc

File: modules/k8s-nmstate-creating-interfaces.adoc [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+
+[id="creating-interfaces_{context}"]
+
+= Creating interfaces
+
+The `kubernetes-nmstate` operator requires the `nmstate.io` API version and the kind of policy, which is `NodeNetworkConfigurationPolicy`. Additionally, each policy has a name as defined by `metadata.name` and a desired state as defined by `spec.desiredState`. The policy's desired state contains a declarative specification of the node's network configuration.
```
Comment (ahardin-rh) [inline modules/k8s-nmstate-creating-interfaces.adoc:9]: s/operator/Operator

File: modules/k8s-nmstate-creating-interfaces.adoc [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+
+The `spec.desiredState.interfaces` section begins the `nmstate` policy. Creating the policy is identical to creating a policy with `nmstate`. Refer to link:nmstate.io[NMState] documentation for additional details on creating `nmstate` policies.
```
Comment (ahardin-rh) [inline modules/k8s-nmstate-creating-interfaces.adoc:11]: Refer to the

File: modules/k8s-nmstate-creating-interfaces.adoc [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+    message: 2/2 nodes successfully configured
+    reason: SuccessfullyConfigured
+    status: "True"
+    type: Available
+----
++
+[NOTE]
+.Status Conditions
+====
+The `status.conditions` section describes the condition of the applied policy. In the foregoing example, the `reason` attribute of the status reports is **SuccessfullyConfigured**, indicating that the nodes received and applied the policy. The condition's `type` attribute is **Available** indicating the configuration is working properly. The `type` attribute can also report **Degraded** indicating a failed state.
```
Comment (ahardin-rh) [inline modules/k8s-nmstate-creating-interfaces.adoc:100]: s/**SuccessfullyConfigured**/`SuccessfullyConfigured`
s/**Degraded**/`Degraded`

File: modules/k8s-nmstate-installing-the-kubernetes-nmstate-operator.adoc [hunk 5]
```diff
@@ -0,0 +1,40 @@
+// This is included in the following assemblies:
+//
+// k8s-nmstate-using-kubernetes-nmstate.adoc
```
Comment (ahardin-rh) [inline modules/k8s-nmstate-installing-the-kubernetes-nmstate-operator.adoc:3]: List the full path to the assembly.

File: modules/k8s-nmstate-installing-the-kubernetes-nmstate-operator.adoc [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+[id="installing-the-kubernetes-nmstate-operator_{context}"]
+
+= Installing the Kubernetes NMState operator
+
+You must install the `kubnetes-nmstate` operator from the OpenShift web console while logged in with administrator privileges. Once installed, the operator can deploy the NMState State Controller as a daemonset across all of the cluster nodes.
```
Comment (ahardin-rh) [inline modules/k8s-nmstate-installing-the-kubernetes-nmstate-operator.adoc:9]: s/operator/Operator

File: modules/k8s-nmstate-installing-the-kubernetes-nmstate-operator.adoc [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+
+.Procedure
+
+. Select **Operators > OperatorHub**.
```
Comment (ahardin-rh) [inline modules/k8s-nmstate-installing-the-kubernetes-nmstate-operator.adoc:13]: *Operators* -> *OperatorHub* per https://github.com/openshift/openshift-docs/blob/master/contributing_to_docs/doc_guidelines.adoc#quick-markup-reference

File: modules/k8s-nmstate-installing-the-kubernetes-nmstate-operator.adoc [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+
+. In the search field below *All Items*, enter `nmstate` and click *Enter*  to search for the `kubernetes-nmstate` operator.
```
Comment (ahardin-rh) [inline modules/k8s-nmstate-installing-the-kubernetes-nmstate-operator.adoc:15]: s/operator/Operator

File: modules/k8s-nmstate-installing-the-kubernetes-nmstate-operator.adoc [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+
+. Click on the `kubernetes-nmstate` operator search result.
+
+. Click on the *Install* button to open the *Install Operator* window.
+
+. Under *Installed Namespace*, ensure the namespace is `openshift-nmstate`. If `openshift-nmstate` does not exist in the combo box, click on *Create Namespace* and enter `openshift-nmstate` in the *Name* field of the dialog box and press the *Create* button.
+
+. Click the *Install* button to install the operator.
```
Comment (ahardin-rh) [inline modules/k8s-nmstate-installing-the-kubernetes-nmstate-operator.adoc:23]: s/operator/Operator

File: modules/k8s-nmstate-installing-the-kubernetes-nmstate-operator.adoc [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+
+. Once the operator finishes installing, click the *View Operator* button.
+
+. Under *Provided APIs*, click *Create Instance* to open the dialog box for creating an instance of `kubernetes-nmstate`.
+
+. In the *Name* field of the dialog box, ensure the name of the instance is `nmstate.`
++
+[NOTE]
+====
+The name restriction is known issue. The instance is a singleton for the entire cluster.
```
Comment (ahardin-rh) [inline modules/k8s-nmstate-installing-the-kubernetes-nmstate-operator.adoc:33]: is a known issue

File: modules/k8s-nmstate-installing-the-kubernetes-nmstate-operator.adoc [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+====
+
+. Accept the default settings and click the *Create* button to create the instance.
+
+.Summary
+
+Once complete, the operator has deployed the NMState State Controller as a daemonset across all of the cluster nodes.
```
Comment (ahardin-rh) [inline modules/k8s-nmstate-installing-the-kubernetes-nmstate-operator.adoc:40]: Operator

File: modules/k8s-nmstate-reading-the-state-of-a-node.adoc [hunk 12]
```diff
@@ -0,0 +1,130 @@
+// This is included in the following assemblies:
+//
+// virt-k8s-nmstate-reporting.adoc
```
Comment (ahardin-rh) [inline modules/k8s-nmstate-reading-the-state-of-a-node.adoc:3]: List the full path to the assembly.

File: modules/k8s-nmstate-reading-the-state-of-a-node.adoc [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+.Procedure
+
+. To return the state of a particular node, specify the node name.
++
+[source,bash]
+----
+[kni@node ~]$ oc get nodenetworkstates <node-name> -o yaml
+----
++
+Replace `<node-name>` with the hostname of the node.
```
Comment (ahardin-rh) [inline modules/k8s-nmstate-reading-the-state-of-a-node.adoc:20]: s/hostname/host name   per IBM Style Guide

File: modules/k8s-nmstate-reading-the-state-of-a-node.adoc [hunk 14]
```diff
@@ -0,0 +1,130 @@
...
+      - destination: ff00::/8
+        metric: 256
+        next-hop-address: ""
+        next-hop-interface: eth0
+        table-id: 255
+      # output truncated
+  lastSuccessfulUpdateTime: "2020-01-31T12:14:00Z"
+----
++
+The object name is the short hostname of the node it represents. The `status.currentState` section contains the node's DNS configuration, list of interfaces, and their configuration and routes.
```
Comment (ahardin-rh) [inline modules/k8s-nmstate-reading-the-state-of-a-node.adoc:125]: host name

File: modules/k8s-nmstate-removing-interfaces.adoc [hunk 15]
```diff
@@ -0,0 +1,61 @@
+// This is included in the following assemblies:
+//
+// k8s-nmstate-configuring.adoc
```
Comment (ahardin-rh) [inline modules/k8s-nmstate-removing-interfaces.adoc:3]: List the full path to the assembly.

File: modules/k8s-nmstate-removing-interfaces.adoc [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+
+[id="removing-interfaces_{context}"]
+
+= Removing interfaces
+
+To remove a configured interface from the cluster's nodes, you must explicitly change the policy from `state: up` to `state: absent` and apply the policy.
+
+[IMPORTANT]
+.Removing a policy doesn't revert the state
```
Comment (ahardin-rh) [inline modules/k8s-nmstate-removing-interfaces.adoc:12]: s/doesn't/does not
We use single asterisks to embolden items, not double asterisks per https://github.com/openshift/openshift-docs/blob/master/contributing_to_docs/doc_guidelines.adoc#quick-markup-reference

I suggest putting "Removing a policy does not revert the state." inside the text box, not in that heading format

File: modules/k8s-nmstate-removing-interfaces.adoc [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+metadata:
+  name: bond0-eth1-eth2
+spec:
+  desiredState:
+    interfaces:
+    - name: bond0
+      state: absent <1>
+----
++
+<1> Change the state from `up` to `absent`
```
Comment (ahardin-rh) [inline modules/k8s-nmstate-removing-interfaces.adoc:34]: `absent`.  (missing period)

File: modules/k8s-nmstate-removing-interfaces.adoc [hunk 18]
```diff
@@ -0,0 +1,61 @@
...
++
+[TIP]
+====
+Save the modified policy with a different file name. The modified policy uses the `bond0-eth1-eth2_absent.yaml` filename to distinguish it from the original `bond0-eth1-eth2.yaml` filename. The policy name in the policy must remain identical.
```
Comment (ahardin-rh) [inline modules/k8s-nmstate-removing-interfaces.adoc:38]: s/filename/file name

File: modules/k8s-nmstate-restoring-the-original-configuration.adoc [hunk 19]
```diff
@@ -0,0 +1,42 @@
+// This is included in the following assemblies:
+//
+// k8s-nmstate-configuring.adoc
```
Comment (ahardin-rh) [inline modules/k8s-nmstate-restoring-the-original-configuration.adoc:3]: List the full path to the assembly

File: modules/k8s-nmstate-restoring-the-original-configuration.adoc [hunk 20]
```diff
@@ -0,0 +1,42 @@
...
+
+[id="restoring-the-original-configuration_{context}"]
+
+= Restoring the original configuration
+
+Removing an interface configuration **does not** restore the network interfaces to the state they were in before a policy was applied. Instead, the NICs targeted in the policy are down and reflect `state: down` when getting the `NodeNetworkState`. For example:
```
Comment (ahardin-rh) [inline modules/k8s-nmstate-restoring-the-original-configuration.adoc:9]: single asterisks to embolden text

File: networking/k8s_nmstate/k8s-nmstate-configuring.adoc [hunk 21]
```diff
@@ -0,0 +1,28 @@
+[id="k8s-nmstate-configuring"]
+
+= Configuring
```
Comment (ahardin-rh) [inline networking/k8s_nmstate/k8s-nmstate-configuring.adoc:3]: I suggest renaming this to be more specific than just "Configuring"

File: networking/k8s_nmstate/k8s-nmstate-configuring.adoc [hunk 22]
```diff
@@ -0,0 +1,28 @@
...
+include::modules/common-attributes.adoc[]
+:context: k8s-nmstate-configuring
+
+The `kubernetes-nmstate` operator enables users to configure network interfaces, DNS settings and routing on {product-title} cluster nodes.
```
Comment (ahardin-rh) [inline networking/k8s_nmstate/k8s-nmstate-configuring.adoc:7]: s/operator/Operator
settings, and routing

File: networking/k8s_nmstate/k8s-nmstate-configuring.adoc [hunk 23]
```diff
@@ -0,0 +1,28 @@
...
+
+Two object types drive the configuration:
+
+- `NodeNetworkConfigurationPolicy`: This object defines the policy, which is the desired network configuration on cluster nodes.
+
+- `NodeNetworkConfigurationEnactment`: This object is read-only and reflects the execution state of a policy on cluster nodes.
+
+The `kubernetes-nmstate` operator can support multiple policies, but you **must** avoid inter-policy dependencies. For example, if you configure policies such that they must be applied in a specific order, the policies must be consolidated into a single policy.
```
Comment (ahardin-rh) [inline networking/k8s_nmstate/k8s-nmstate-configuring.adoc:15]: Operator
remove emphasis around must

File: networking/k8s_nmstate/k8s-nmstate-using-kubernetes-nmstate.adoc [hunk 24]
```diff
@@ -0,0 +1,20 @@
+[id="k8s-nmstate-using-kubernetes-nmstate"]
+
+= Using Kubernetes NMState
+include::modules/common-attributes.adoc[]
+:context: k8s-nmstate-using
+
+The `kubernetes-nmstate` operator provides a Kubernetes API for peforming state-driven network configuration across the {product-title} cluster's nodes with `nmstate`. `kubernetes-nmstate` provides users with functionality to configure various network interface types, DNS and routing on cluster nodes. Additionally, the daemons on the cluster nodes periodically report on the state of each node's network interfaces to the API server.
```
Comment (ahardin-rh) [inline networking/k8s_nmstate/k8s-nmstate-using-kubernetes-nmstate.adoc:7]: Operator

File: networking/k8s_nmstate/k8s-nmstate-using-kubernetes-nmstate.adoc [hunk 25]
```diff
@@ -0,0 +1,20 @@
...
+
+
+[NOTE]
+.Technology Preview
+====
+The `kubernetes-nmstate` operator is a Technology Preview feature only. Technology Preview features are not supported with Red Hat production service level agreements (SLAs) and might not be functionally complete. Red Hat does not recommend using them in production. These features provide early access to upcoming product features, enabling customers to test functionality and provide feedback during the development process.
+
+For more information about the support scope of Red Hat Technology Preview features, see link:https://access.redhat.com/support/offerings/techpreview/[Technology Preview Features Support Scope].
+
+`nmstate` requires Network Manager installed on each cluster node.
```
Comment (ahardin-rh) [inline networking/k8s_nmstate/k8s-nmstate-using-kubernetes-nmstate.adoc:17]: I see both Network Manager and NetworkManager usage, so we just need to be consistent.

File: networking/k8s_nmstate/k8s-nmstate-using-kubernetes-nmstate.adoc [hunk 26]
```diff
@@ -0,0 +1,20 @@
+[id="k8s-nmstate-using-kubernetes-nmstate"]
+
+= Using Kubernetes NMState
+include::modules/common-attributes.adoc[]
+:context: k8s-nmstate-using
+
+The `kubernetes-nmstate` operator provides a Kubernetes API for peforming state-driven network configuration across the {product-title} cluster's nodes with `nmstate`. `kubernetes-nmstate` provides users with functionality to configure various network interface types, DNS and routing on cluster nodes. Additionally, the daemons on the cluster nodes periodically report on the state of each node's network interfaces to the API server.
+
+
+[NOTE]
```
Comment (ahardin-rh) [inline networking/k8s_nmstate/k8s-nmstate-using-kubernetes-nmstate.adoc:10]: There is a module called technology-preview.adoc that you could use here instead

File: networking/k8s_nmstate/k8s-nmstate-using-kubernetes-nmstate.adoc [hunk 27]
```diff
@@ -0,0 +1,20 @@
...
+.Technology Preview
+====
```
Comment (ahardin-rh) [inline networking/k8s_nmstate/k8s-nmstate-using-kubernetes-nmstate.adoc:12]: I would remove the .Technology Preview heading

File: networking/k8s_nmstate/k8s-nmstate-using-kubernetes-nmstate.adoc [hunk 28]
```diff
@@ -0,0 +1,20 @@
...
+The `kubernetes-nmstate` operator is a Technology Preview feature only. Technology Preview features are not supported with Red Hat production service level agreements (SLAs) and might not be functionally complete. Red Hat does not recommend using them in production. These features provide early access to upcoming product features, enabling customers to test functionality and provide feedback during the development process.
```
Comment (ahardin-rh) [inline networking/k8s_nmstate/k8s-nmstate-using-kubernetes-nmstate.adoc:13]: Operator

File: modules/k8s-nmstate-creating-interfaces.adoc [hunk 29]
```diff
@@ -0,0 +1,250 @@
...
+         - eth2
+----
++
+The foregoing exemplary YAML content is a policy manifest for bonding NICs `eth1` and `eth2` on the {product-title} cluster's nodes in a YAML file named `bond0-eth1-eth2_up.yaml`.
+
+. Apply the policy manifest.
++
+[source,bash]
+----
+[kni@node ~]$ oc apply -f bond0-eth1-eth2_up.yaml
```
Comment (ahardin-rh) [inline modules/k8s-nmstate-creating-interfaces.adoc:46]: For each of the commands throughout this PR, do we need the preceding [kni@node ~] ?

File: networking/k8s_nmstate/k8s-nmstate-reporting.adoc [hunk 30]
```diff
@@ -0,0 +1,12 @@
+[id="virt-k8s-nmstate-reporting"]
+
+= Reporting
+include::modules/common-attributes.adoc[]
+:context: k8s-nmstate-reporting
+
```
Comment (ahardin-rh) [inline networking/k8s_nmstate/k8s-nmstate-reporting.adoc:6]: Each of the assemblies is missing a TOC 
You can add: 

toc::[]

File: networking/k8s_nmstate/k8s-nmstate-using-kubernetes-nmstate.adoc [hunk 31]
```diff
@@ -0,0 +1,20 @@
+[id="k8s-nmstate-using-kubernetes-nmstate"]
+
+= Using Kubernetes NMState
+include::modules/common-attributes.adoc[]
+:context: k8s-nmstate-using
+
```
Comment (ahardin-rh) [inline networking/k8s_nmstate/k8s-nmstate-using-kubernetes-nmstate.adoc:6]: toc::[]

File: networking/k8s_nmstate/k8s-nmstate-configuring.adoc [hunk 32]
```diff
@@ -0,0 +1,28 @@
+[id="k8s-nmstate-configuring"]
+
+= Configuring
+include::modules/common-attributes.adoc[]
+:context: k8s-nmstate-configuring
+
```
Comment (ahardin-rh) [inline networking/k8s_nmstate/k8s-nmstate-configuring.adoc:6]: toc::[]

File: _topic_map.yml [hunk 33]
```diff
@@ -813,6 +813,16 @@ Topics:
   - Name: Configuring ingress cluster traffic using a NodePort
     File: configuring-ingress-cluster-traffic-nodeport
     Distros: openshift-enterprise,openshift-webscale,openshift-origin
+# Kubernetes NMState (TECHNOLOGY PREVIEW)
+- Name: Kubernetes NMState (Tech Preview)
+  Dir: k8s_nmstate
+  Topics:
+  - Name: Using Kubernetes NMState
+    File: k8s-nmstate-using-kubernetes-nmstate
+  - Name: Reporting
```
Comment (ahardin-rh) [inline _topic_map.yml:822]: Reporting what? I suggest making the title stronger

File: _topic_map.yml [hunk 34]
```diff
@@ -813,6 +813,16 @@ Topics:
...
+    File: k8s-nmstate-reporting
+  - Name: Configuration
```
Comment (ahardin-rh) [inline _topic_map.yml:824]: Same here. We need to be more specific about what is being configured.

File: _topic_map.yml [hunk 35]
```diff
@@ -813,6 +813,16 @@ Topics:
   - Name: Configuring ingress cluster traffic using a NodePort
     File: configuring-ingress-cluster-traffic-nodeport
     Distros: openshift-enterprise,openshift-webscale,openshift-origin
+# Kubernetes NMState (TECHNOLOGY PREVIEW)
+- Name: Kubernetes NMState (Tech Preview)
```
Comment (ahardin-rh) [inline _topic_map.yml:817]: I don't think we call out Tech Preview in navigation. We use the technology-preview.adoc to consistently state what is in Technology Preview.

**Comment by ahardin-rh** [comment]
@johnwilkins Overall, looks great! Just some miscellaneous style comments from me.

File: networking/k8s_nmstate/k8s-nmstate-configuring.adoc [hunk 36]
```diff
@@ -0,0 +1,28 @@
...
+The `kubernetes-nmstate` operator enables users to configure network interfaces, DNS settings and routing on {product-title} cluster nodes.
+
+Two object types drive the configuration:
+
+- `NodeNetworkConfigurationPolicy`: This object defines the policy, which is the desired network configuration on cluster nodes.
```
Comment (jboxman) [inline networking/k8s_nmstate/k8s-nmstate-configuring.adoc:11]: We aren't using _desired_.

>desired adjective
Do not use. Use "required" or "the that you want".

Maybe for k8s _intended_ or _target configuration_ is better?

File: networking/k8s_nmstate/k8s-nmstate-reporting.adoc [hunk 37]
```diff
@@ -0,0 +1,12 @@
...
+Once installed, each node in the cluster contains a `NodeNetworkSState` object. The daemonset periodically reports the state of each object to the API server.
```
Comment (jboxman) [inline networking/k8s_nmstate/k8s-nmstate-reporting.adoc:7]: *NodeNetworkState*

Comment (jboxman) [inline networking/k8s_nmstate/k8s-nmstate-reporting.adoc:7]: s/daemonset/daemon set/

File: modules/k8s-nmstate-creating-interfaces.adoc [hunk 38]
```diff
@@ -0,0 +1,250 @@
...
+  - lastHearbeatTime: "2020-02-07T10:27:09Z"
+    lastTransitionTime: "2020-02-07T10:27:09Z"
+    message: 2/2 nodes successfully configured
+    reason: SuccessfullyConfigured
+    status: "True"
+    type: Available
+----
++
+[NOTE]
+.Status Conditions
```
Comment (jboxman) [inline modules/k8s-nmstate-creating-interfaces.adoc:98]: s/Conditions/conditions/

**Review by jboxman — CHANGES_REQUESTED** [review]
So I made a first pass on this, and left comments for some items I discovered for your review. Some apply in multiple places throughout, but for brevity I only tagged a specific instance.

Thanks!

**Comment by jboxman** [comment]
@johnwilkins, I'm unsure about the topics for this. _Using_ is just installing the Operator.

Configuring is both creating and removing interfaces. Is this a single task? Will a user be doing both of these as part of the same user story?

Also, some of this content largely duplicates existing content under `modules/virt-*` as near as I can tell. Is it possible to reuse existing modules, rather than have multiple, similar modules?

File: _topic_map.yml (hunk 35 above)
[the remark this answers is not in the document]
Comment (johnwilkins) [reply _topic_map.yml:817]: Apparently so. https://www.nmstate.io/kubernetes-nmstate/

**Comment by ahardin-rh** [comment]
Closing in favor of https://github.com/openshift/openshift-docs/pull/29400

**Pull request closed without merging**